### PR TITLE
Some performance improvements

### PR DESCRIPTION
**Pull request by gluck**
Each commit implements one.
Some more to come eventually.

Thx.

### CODE REVIEW

File: Mono.Cecil/Import.cs [hunk 1]
```diff
@@ -759,7 +759,7 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
 
 			for (int i = 0; i < references.Count; i++) {
 				var reference = references [i];
-				if (name_reference.FullName != reference.FullName) // TODO compare field by field
+				if (!Equals(name_reference, reference))
```
Comment (jbevain) [inline Mono.Cecil/Import.cs]: `!Equals (`

File: Mono.Cecil/Import.cs [hunk 2]
```diff
@@ -769,6 +769,32 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
 			assembly_reference = null;
 			return false;
 		}
+
+		private static bool Equals(byte[] a, byte[] b)
```
Comment (jbevain) [inline Mono.Cecil/Import.cs]: `Equals (byte [] a, byte [] b)`

File: Mono.Cecil/Import.cs [hunk 3]
```diff
@@ -769,6 +769,32 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
...
+		{
+			if (a == b) return true;
```
Comment (jbevain) [inline Mono.Cecil/Import.cs]: Object.ReferenceEquals. Also put the return on a new line.

File: Mono.Cecil/Import.cs [hunk 4]
```diff
@@ -769,6 +769,32 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
...
+			if (a.Length != b.Length) return false;
+			for (int i = 0; i < a.Length; i++)
+				if (a[i] != b[i]) return false;
```
Comment (jbevain) [inline Mono.Cecil/Import.cs]: `a [i] != b [i]`

File: Mono.Cecil/Import.cs [hunk 5]
```diff
@@ -769,6 +769,32 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
...
+			return true;
+		}
+
+		private static bool Equals<T>(T a, T b) where T : class, IEquatable<T>
```
Comment (jbevain) [inline Mono.Cecil/Import.cs]: `Equals<T> (T a, T b)`

File: Mono.Cecil/Import.cs [hunk 6]
```diff
@@ -769,6 +769,32 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
...
+		{
+			if (a == b) return true;
+			if (a == null) return false;
+			return a.Equals(b);
```
Comment (jbevain) [inline Mono.Cecil/Import.cs]: `a.Equals (b)`

File: Mono.Cecil/Import.cs [hunk 7]
```diff
@@ -769,6 +769,42 @@ public static bool TryGetAssemblyNameReference (this ModuleDefinition module, As
 			assembly_reference = null;
 			return false;
 		}
+
+		private static bool Equals (byte [] a, byte [] b)
+		{
+			if (Object.ReferenceEquals (a, b))
+				return true;
+			if (a.Length != b.Length)
```
Comment (gluck) [inline Mono.Cecil/Import.cs]: I should check for null here (PublicKeyToken is never null today, but that may change or this helper could be reused)

Comment (jbevain) [inline Mono.Cecil/Import.cs]: Agreed.

**Comment by jbevain** [comment]
Can you squash the last formatting commits into the original before we merge this? Thanks!

**Comment by gluck** [comment]
Done

**Comment by jbevain** [comment]
Thanks!

**Pull request merged**